### PR TITLE
Table Panel: Add ability to set background colors for entire rows

### LINKED ISSUE #11418: [Feat] Conditional formatting
## Introduction

As of writing (Grafana v5.1-pre) we have support for 
* coloring background/value in the Singlestat panel based on a numeric range and 3 colors
* coloring cell/value/row in the Table panel based on a numeric range and 3 colors

image

image

There are a lot of feature requests/issues requesting greater support of coloring in table panel and the singlestat panel:
* #7631 [Feature request] Allow Text to Value mappings in Singlestat panel
* #3601 table panel: colorize cell value based on regex pattern match
* #9500 Feature request: Allow thresholds for Time column in tables
* #3608 Thresholds with more than 3 comma-separated values
* #8404 [Feature request] Singlestat null value color 
* #4821 Table Panel coloring shouldn't require display of referenced column
* #9709 Single stat panel with gradient color
* #7411 [Feature request] Variable threshold values in Singlestats panel
* #7286 [Feature request] Row and/or Column styles in Table

Historically we've received a few PR's that tried to solve some of the issues above, but we haven't been able to merge them because we feel the solutions are either hard to understand from a user perspective or either it makes the Grafana code base/features more and more shattered.

## Proposal

This issue aims to find one solution for solving all (or most) of the issues above and by that delivering a good and consistent user experience. By that said, we probably want to split this up in smaller issues if we decide that this is a good way forward.

In general the *Conditions editor* for alerts are very similar to a rule/conditional formatting editor functionality and would probably fit quite nice with a few modifications. This is something to take into considerations when implementing this.

image

### Conditional rules evaluator

A panel needing conditional formatting shouldn't need to implement their own solution for evaluating conditional rules thus making this a core component/helper library is a must. 
It would be nice if this component also could return a list of rule types that it supports.

Examples of rule types:
* Value is empty
* Value is not empty
* Text contains
* Text does not contain
* Text starts with
* Text ends with
* Text is exactly
* Date is
* Date is before
* Date is after
* Greater than
* Greater than or equal to
* Less than
* Less than or equal to
* Is equal to
* Is not equal to
* Is between
* Is not between

### Singlestat panel - colorize background/value

Changes here compared to current functionality is that 
* it allow you to define unlimited amount of color formatting's compared to today's maximum of three.
* it allow support for coloring based on empty/non-empty values
* it allow support for coloring based on string/text values
* it allow support for coloring based on date/duration values

#### UX mockup/example to underline functionality needed/wanted

image

### Table panel - colorize cells/values/rows

Changes here compared to current functionality is that 
* it allow you to define unlimited amount of color formatting's compared to today's maximum of three.
* it allow support for coloring based on empty/non-empty values
* it allow support for coloring based on string/text values
* it allow support for coloring based on date/duration values
* it allow support for coloring a row based on a hidden/non-rendered column

Currently you can define thresholds for each defined *Column style*. Maybe we want to continue having this possibility since we can make smart decisions of what columns to apply the conditional formatting rules to. 

On the other hand it quite weird to continue using the column styles since they're basically provides options for how to render column cells/convert raw column value using specific data types and options etc. Regarding coloring we currently supports cell/value/row which affects cells or rows and rules are evaluated based on  the raw data returned from data source query, that is not after conversion/options. In the mockup below there's a really simple UX which basically would scan all rows and columns returned from data source query to apply conditional formatting.

One option to having conditional formatting inside the *Column style* tab would be to allow conditional formatting rules to be applied per columns basis under a specific tab. Basically like the mockup below but with additional logic to decide which columns the rules should be applied to. This solution have an advantage because it can support conditional formatting of rows based on a hidden/non-rendered column.

#### UX mockup/example to underline functionality needed/wanted

image

### Forward compatibility

Any changes needed to existing Singlestat and Table panel JSON models should be forward compatible, meaning that existing saved dashboards with Singlestat/Table panel having coloring/threshold enabled should be migrated/converted to the new solution when such a dashboard is rendered.

### PR DESCRIPTION
**Pull request by codeincarnate**
Riffing on #83247, added support for the toggling the setting. Supports gradients and solid colors. Also updated the hover behavior so that it works with colors better. Might be worth elaborating there, but can also save for the future potentially 😄. Think it should also be straight-forward to support this with the colored text mode as well. 

Fixes https://github.com/grafana/grafana/issues/11418

https://github.com/grafana/grafana/assets/199847/175b0393-e689-4dd5-9c74-9ee8c303e21d

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by gtk-grafana** [comment]
Seeing some weirdness with value mappings when the new option is enabled:
image
image
Look at the second screenshot and then check the value mappings, looks like the apply to entire row is causing the value mappings to shift around.

**Comment by codeincarnate** [comment]
Whoa strange, not really sure what's going there but will take a look 👀

**Comment by codeincarnate** [comment]
Looks like sorting is messing as things are getting sorted, but I assume keeping the same index in the underlying data being used for coloring. I'll look at getting a fix for this tomorrow

File: packages/grafana-ui/src/components/Table/RowsList.tsx [hunk 1]
```diff
@@ -202,16 +201,23 @@ export const RowsList = (props: RowsListProps) => {
...
-      rowBg = (rowIndex: number) => {
-        const display = field.display!(field.values.get(rowIndex));
-        return display.color;
+    if (
+        fieldOptions.cellOptions?.type === TableCellDisplayMode.ColorBackground &&
+        fieldOptions.cellOptions.applyToRow
+    ) {
+      const isGradient =
+        fieldOptions.cellOptions.mode === undefined ||
+        TableCellBackgroundDisplayMode.Gradient === fieldOptions.cellOptions.mode;
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Table/RowsList.tsx:215]: not sure why we care about this here, we can just always set style.background

File: packages/grafana-ui/src/components/Table/RowsList.tsx [hunk 2]
```diff
@@ -234,8 +239,15 @@ export const RowsList = (props: RowsListProps) => {
...
-        style.color = getTextColorForAlphaBackground(style.backgroundColor!, tableStyles.theme.isDark);
+        const { colors, isGradient } = rowBg(index);
+        style.color = colors.textColor;
+        
+        if (isGradient) {
+          style.backgroundImage = colors.bgColor;
+        }
+        else {
+          style.backgroundColor = colors.bgColor;
+        }
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Table/RowsList.tsx:250]:
```suggestion
        const colors = rowBg(index);
        style.color = colors.textColor;
        style.background = colors.bgColor;
```

File: packages/grafana-ui/src/components/Table/DefaultCell.tsx [hunk 3]
```diff
@@ -101,28 +100,17 @@ function getCellStyle(
...
-      textColor = getTextColorForAlphaBackground(displayValue.color!, tableStyles.theme.isDark);
-      bgColor = `linear-gradient(120deg, ${bgColor2.toRgbString()}, ${displayValue.color})`;
-    }
+  // Get colors
+  const colors = getCellColors(tableStyles, cellOptions, displayValue);
+  textColor = colors.textColor;
+  
+  // Skip application of background color if we're applying to the whole row
+  if (cellOptions.type === TableCellDisplayMode.ColorBackground && !cellOptions.applyToRow) {
+    bgColor = colors.bgColor;
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Table/DefaultCell.tsx:113]: this is a bit of shame as it means we cannot override the color for specific fields. I guess we need to disable this to make the gradient not look strange (broken up)?

**Review by torkelo — COMMENTED** [review]
Looks good, but can be simplifed a bit

File: packages/grafana-ui/src/components/Table/DefaultCell.tsx (hunk 3 above)
Comment (codeincarnate) [reply packages/grafana-ui/src/components/Table/DefaultCell.tsx:113]: Yep, this is to have the gradient stretch across the row as opposed to the row and the cell as well. Separate field overrides might still work. I can take a look there 😃

File: packages/grafana-ui/src/components/Table/RowsList.tsx (hunk 2 above)
Comment (codeincarnate) [reply packages/grafana-ui/src/components/Table/RowsList.tsx]: Went for using style.background, definitely more simple 😄

**Comment by codeincarnate** [comment]
@gtk-grafana pretty sure I have the bug with sorting fixed if you want to check this again

**Comment by gtk-grafana** [comment]
Nice, yeah I'm not seeing that issue anymore, but I did run across one thing with the font colors:

image
image

image

It looks as though the cell that is being value matched updates the font color to show a light font on dark backgrounds and the other way around, but the other cells in the row do not, and you can get some rows that are really difficult to read.

**Comment by gtk-grafana** [comment]
Oh, and also if you turn on inspect, the cell with the correct font color seems to get incorrectly applied to the inspect icon, unless the inspect icon isn't supposed to have a background: 
image
image

**Comment by codeincarnate** [comment]
Thanks for taking a look again Galen, this is great feedback! 🙏  Will get further fixes in 😄

**Comment by codeincarnate** [comment]
Fixed the text issue. Also updated the treatment for the `Cell Inspect` button. I think it helps but would be curious about opinions on that one.

**Comment by gtk-grafana** [comment]
looks good and is working as expected locally. Inspect looks much better and seems to match the font color now.

Until I merge in main: then it looks like the row background color conflicts with the new overflow styles:
image
image

Also there are some tests failing in drone

**Comment by codeincarnate** [comment]
I'll fix up those tests and get main merged in 😄. I suppose in that case it would be best to merge into main instead so I'll see what I can do there 👍

**Comment by gtk-grafana** [comment]
Hey, one more thing, although this appears like a problem as well with the normal colored backgrounds not applied to the row as well: the hover styles of the row are being removed when inspect is turned on, and when inspect is turned off, the hover style is only applied on the cell the user is hovering over

**Comment by codeincarnate** [comment]
> the hover styles of the row are being removed when inspect is turned on, and when inspect is turned off, the hover style is only applied on the cell the user is hovering over

Oooh interesting, seems like the code for inspect is a bit tangled. I'll try to get this all untangled today 😄

**Comment by codeincarnate** [comment]
Oof merging in main made this a beast, probably going to do this relative to main instead. Got the hovering behavior mostly fixed though I'm running into an issue when things aren't colored. Taking longer than expected but hoping to have it 100% done tomorrow

**Comment by codeincarnate** [comment]
Changed the base branch as the merge from main was rough. @gtk-grafana I believe I've fixed the color problems completely at this point

**Comment by codeincarnate** [comment]
@gtk-grafana gave the cell value inspect another pass. Using a dark background with some alpha to still show colors. If you could take a look it would be appreciated 🙏 

Otherwise, I'm running into an issue as currently field color by default is being set to `#00000000` i.e. a fully transparent black. I had been setting this non-transparent, but in light mode it makes the table look dark by default. Pretty undeseriable and I'm currently figuring out where this is being set.

**Comment by gtk-grafana** [comment]
@codeincarnate 
Yeah, agreed the fix works great for dark backgrounds
image
image

But yeah on light backgrounds I'm seeing that issue we saw earlier where the inspect icon color is only appropriate on the field that is being matched
image
image
image

**Comment by codeincarnate** [comment]
Ok, I'm closer. Turns out there is a bug in the field display processor in this case. Here's a video from Grafana 10.2.3: 


https://github.com/grafana/grafana/assets/199847/b03f712c-ae8b-4f5a-80f0-ee442b188c82

Turns out at least in light mode the only color coming back is transparent black so nothing is displayed 😢. I think style overrides are saving us in dark mode but will try to address that here as well

**Comment by codeincarnate** [comment]
Ah, didn't realize that the base threshold for that table was set to transparent 😆. That would do it, that probably needs a little improvement but I think we don't need to do anything particularly to handle base color threshold.

**Comment by gtk-grafana** [comment]
@codeincarnate I'm not sure I understand so I'll shotgun more screenshots :P It looks like the color base is changing the behavior?

Dark color base with light matched row
image
image

Light color base with dark matched row
image
image

Text color as base:
image
image
image
image

**Comment by gtk-grafana** [comment]
These are getting pretty nitpicky/edge-case, if it's a big lift to fix these I could be convinced that we should merge as is

**Comment by codeincarnate** [comment]
No worries @gtk-grafana, I can do a couple of fixes today and that should hopefully be sufficient 😄

**Review by gtk-grafana — APPROVED** [review]
Works as expected locally!

**Pull request merged**